mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -190,10 +190,11 @@ def _verify_listed_object_contains_artifact_path_prefix(listed_object_path, arti
             )
 
     def _download_file(self, remote_file_path, local_path):
-        (bucket, s3_root_path) = data.parse_s3_uri(self.artifact_uri)
-        s3_full_path = posixpath.join(s3_root_path, remote_file_path)
-        s3_client = self._get_s3_client()
-        s3_client.download_file(bucket, s3_full_path, local_path)
```

Comment:
@michaguenther Does `s3_client.download_file` overwrite the existing file?

Reply:
@michaguenther How can we verify that the local file is up-to-date?

Reply:
@dbczumar @harupy I have further developed the check. In case a file already exists locally I get the etag, hash of the file in the s3 bucket (without downloading the file in the first place). Then I compare it with the local hash and skip the download if the hashes match. The hash comparison was adopted from [stackoverflow](https://stackoverflow.com/questions/1775816/how-to-get-the-md5sum-of-a-file-on-amazons-s3).

Reply:
@michaguenther Thanks! Is there an S3 client API for computing this ETag? Alternatively, does AWS provide official code for calculating it? Reason I ask is that the etag doesn't always seem to be an md5 hash: https://stackoverflow.com/a/41735844/11952869.

Reply:
@dbczumar There seems to be a [new api](https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/), but this has to be activated in the aws settings. 
The code be then further optimised with the following [method](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.get_object_attributes).

I am not sure if your described issue can be resolved at the moment. 
Overall the worst thing that happens with the current code is that the file gets downloaded although it would not have been necessary but for many case we manage to skip the download.